src/pcre2.h.generic

```diff
@@ -464,6 +464,17 @@ released, the numbers must not be changed. */
 #define PCRE2_CONFIG_COMPILED_WIDTHS        14
 #define PCRE2_CONFIG_TABLES_LENGTH          15
 
+/* Optimization directives for pcre2_set_optimize().
+For binary compatibility, only add to this list; do not renumber. */
+
+#define PCRE2_OPTIMIZATION_NONE   0
+#define PCRE2_OPTIMIZATION_FULL   1
+#define PCRE2_DO_AUTO_POSSESS     2
+/* Do not use 3 */
+#define PCRE2_DO_DOTSTAR_ANCHOR   4
+/* Do not use 5 */
+#define PCRE2_DO_START_OPTIMIZE   6
+/* Do not use 7 */
```

Comment:
I try to understand the proposed mechanic. Each code represents two bitsets. One set is added, the other is removed from the optimization list when specified. The "Do not use number" is removal of a feature.

For simplification, I would give a name to the removal cases as well. Furthermore PCRE2_DO_AUTO_POSSESS should be started from 64, so there is space to add other groups like PCRE2_OPTIMIZATION_FULL later. The bits >= 64 only change one optimization (DO/DONT), they don't need bitsets. The modes < 64 sets the default configuration, they need a single bitset.

Reply:
It doesn't really work that well, but this is a draft for a reason.  the few options that work (ex: PCRE2_OPTIMIZATION_NONE) skip all the compatibility stuff and then look at record 0 in both vectors, so `|=0` and then `&=~7`, resulting in `optimization_flags == 0`., the `PCRE2_DO_*` should work the same but I suspect that they haven't been really tested.

The point is that whatever logic is needed to enable/disable flags could be added to the implementation and since there are 32bit values that can be assigned then there is no longer an API breakage when we run out of bits on that flag.

Reply:
I understand the point, and this is why I suggested the changes above. We have a set which can be used for setting generally useful options (users don't need to know the specific details), and individual optimizations can be controlled one by one. We can have two API functions for them, or add a separator constant (64 suggested).

Reply:
> For simplification, I would give a name to the removal cases as well.

They do have names... unfortunately, the names I want to use are already defined with other values, which don't fit into the contiguous range of integers which are needed to index into the add and remove bitset arrays.

The names I want to use are `PCRE2_NO_AUTO_POSSESS`, `PCRE2_NO_DOTSTAR_ANCHOR`, etc. These are already defined as 0x4000, 0x8000, and 0x10000, for use in the compile options. The values of `PCRE2_NO_AUTO_POSSESS`, etc. can't be changed or that would break backwards compatibility. So in `pcre2_set_optimization` I am converting to the desired indices with this code:

```c
PCRE2_EXP_DEFN int PCRE2_CALL_CONVENTION
pcre2_set_optimize(pcre2_compile_context *ccontext, uint32_t directive)
{
/* We are reusing some symbolic constants which were already defined before
 * pcre2_set_optimize() was added; convert to indexes in conversion tables */
if (directive == PCRE2_NO_AUTO_POSSESS)
  directive = 3;
else if (directive == PCRE2_NO_DOTSTAR_ANCHOR)
  directive = 5;
else if (directive == PCRE2_NO_START_OPTIMIZE)
  directive = 7;
```

Do you think it will likely cause confusion if `pcre2_set_optimization()` re-uses the same constant names which are used for `pcre2_compile()` option flags? If so, can you suggest some unused names which would be better?

Reply:
> I try to understand the proposed mechanic. Each code represents two bitsets. One set is added, the other is removed from the optimization list when specified.

Well, that's the current implementation.

Really, though, from the perspective of external callers, what we have is a 32-bit space of arbitrary integral values, each of which tells PCRE2 *something* about what optimization behavior is desired. You can think of these "optimization directive" numbers kind of like system call numbers used when calling into an OS kernel from userspace; to the caller, it's just an arbitrarily defined number.

Internally, PCRE2 can store the optimization settings in any convenient manner. Right now it happens to be a bitset, so using the optimization directive numbers to index into set/clear arrays is convenient. But in the future, it could be anything.

Reply:
> For simplification, I would give a name to the removal cases as well. Furthermore PCRE2_DO_AUTO_POSSESS should be started from 64, so there is space to add other groups like PCRE2_OPTIMIZATION_FULL later. The bits >= 64 only change one optimization (DO/DONT), they don't need bitsets. The modes < 64 sets the default configuration, they need a single bitset.

Fair enough, I think the "two arrays of set/clear bitsets" was pretty elegant but let me try this. Will push an update shortly.

Reply:
OK, please see the commit I have just pushed. Let me know if you prefer this implementation or the previous one.